### PR TITLE
Issue 3108: List all the scopes during listScopes call

### LINKED ISSUE #3108: listScopes returns 500 when auth is enabled
**Problem description**

When auth is enabled on the controller and segment store, listScopes returns 500.  Create scope and get one scope behave normally.

This is using the built-in basic auth handler with password file.

```
Get scopes without credentials, 401 as expected

usxxcbalczunm1:pravega-auth cbalczun$ curl http://...:10080/v1/scopes -H "Content-Type: application/json" -i
HTTP/1.1 401 Unauthorized
Content-Length: 0

Now passing the right credentials:

usxxcbalczunm1:pravega-auth cbalczun$ curl http://...:10080/v1/scopes -H "Authorization: Basic YWRtaW46MTExMV9hYWFh" -H "Content-Type: application/json" -i
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 0

Create a scope

usxxcbalczunm1:pravega-auth cbalczun$ curl http://...:10080/v1/scopes -H "Authorization: Basic YWRtaW46MTExMV9hYWFh" -H "Content-Type: application/json" -i -d "{ \"scopeName\": \"myscope\"}"
HTTP/1.1 201 Created
Content-Type: application/json
Content-Length: 23

{"scopeName":"myscope"}

Get one scope:

usxxcbalczunm1:pravega-auth cbalczun$ curl http://...:10080/v1/scopes/myscope -H "Authorization: Basic YWRtaW46MTExMV9hYWFh" -H "Content-Type: application/json" -i
HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: 23

{"scopeName":"myscope"}

But list all scopes:

usxxcbalczunm1:pravega-auth cbalczun$ curl http://...:10080/v1/scopes -H "Authorization: Basic YWRtaW46MTExMV9hYWFh" -H "Content-Type: application/json" -i
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 0
```

When the 500 occurs, this can be observed in the controller logs:

```
2018-11-15 02:13:02,124 77882 [controllerpool-10] WARN  i.p.c.s.r.r.StreamMetadataResourceImpl - listScopes failed with exception:
java.util.concurrent.CompletionException: java.lang.IllegalStateException: Not inside a request scope.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.pravega.controller.store.stream.ZKStoreHelper.lambda$getChildren$9(ZKStoreHelper.java:155)
	at io.pravega.controller.store.stream.ZKStoreHelper.lambda$callback$23(ZKStoreHelper.java:305)
	at org.apache.curator.framework.imps.Backgrounding$1$1.run(Backgrounding.java:158)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Not inside a request scope.
	at org.glassfish.jersey.internal.guava.Preconditions.checkState(Preconditions.java:169)
	at org.glassfish.jersey.process.internal.RequestScope.current(RequestScope.java:177)
	at org.glassfish.jersey.inject.hk2.RequestContext.findOrCreate(RequestContext.java:79)
	at org.jvnet.hk2.internal.MethodInterceptorImpl.internalInvoke(MethodInterceptorImpl.java:89)
	at org.jvnet.hk2.internal.MethodInterceptorImpl.invoke(MethodInterceptorImpl.java:125)
	at org.jvnet.hk2.internal.MethodInterceptorInvocationHandler.invoke(MethodInterceptorInvocationHandler.java:62)
	at com.sun.proxy.$Proxy56.getRequestHeader(Unknown Source)
	at io.pravega.controller.server.rest.resources.StreamMetadataResourceImpl.authorize(StreamMetadataResourceImpl.java:170)
	at io.pravega.controller.server.rest.resources.StreamMetadataResourceImpl.lambda$null$25(StreamMetadataResourceImpl.java:516)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at io.pravega.controller.server.rest.resources.StreamMetadataResourceImpl.lambda$listScopes$26(StreamMetadataResourceImpl.java:514)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 13 common frames omitted
```

**Problem location**

Controller rest api (I think).

**Suggestions for an improvement**

Return 200 with the list of authorized-to-see scopes.

### PR DESCRIPTION
**Pull request by arvindkandhare**
Signed-off-by: arvindkandhare <arvind.kandhare@emc.com>

**Change log description**  
Removes auth for each scope during `listScopes` call.

**Purpose of the change**  
Fixes #3108.

**What the code does**  

Implementation of auth*n for `listScopes` is as follows:
1. Check whether the principal has access to `/`
2. Check whether the prinicipal has access to individual scopes
3. Return only the scopes that the principal has access to.

This gives wrong info through the UI.  This is also not in sync with other `list*` operations (e.g. `listStreams`). The generic behavior is that if a principal has appropriate access permissions for a container (e.g. `scope` in case of `listStreams`, `\` in case of `listScopes`), listing is allowed.

Updating the behavior of `listScopes` to match other implementations.

**How to verify it**  
Unit tests pass.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 1]
```diff
@@ -512,12 +512,7 @@ public void listScopes(final SecurityContext securityContext, final AsyncRespons
                          .thenApply(scopesList -> {
                              ScopesList scopes = new ScopesList();
                              scopesList.forEach(scope -> {
-                                 try {
-                                     authorize(scope, principal, READ);
```
Comment (sarlaccpit) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: I understand that removing this fixes the issue where someone has READ on / (and therefore shouldn't be required to have READ on individual scopes underneath).

But what if there are 10 scopes, and I have read on scope number 3.   When I do list on /, I should get a list back with just scope number 3 in it, I believe?

With the current code:   I would get access denied because I don't have READ on / 
With the proposes changes:   Same.

(I guess this PR doesn't make things worse and at least fixes the issue for the use case when I have READ on /.   Just pointing out if there's another problem along side of it?)

Comment (sarlaccpit) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: So maybe the ultimate fix is:  look at "/" first,  and if not authorized there, don't bail, look at individual scope permissions?

Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: This is the current implementation. Please see this: https://github.com/pravega/pravega/pull/3110/files/3f482a46aaa92dcf2c1330d82312ebb6049b2044#diff-7d4abaf3b0a8ec424c0bec59a93cf94dL503. ListScopes will list all the scopes in the scenario that you mentioned.

Comment (sarlaccpit) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: But if I don't have READ on /.  Just on scope 3, what happens?

Comment (EronWright) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: My feeling is that `ListScopes` should list all the scopes that one has access to, or an empty list if one has access to no scopes.   Meanwhile, it makes sense that one may be granted access to all scopes, or granted access only to specific scopes.    

@arvindkandhare don't you agree that ListScopes should filter its results?

Comment (sarlaccpit) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: The analogy with the filesystem does make sense.     But as @fpj explains, it's also true, we'll need to be able to get a filtered list of "only what you can see".   So a UI application wouldn't present scopes to a user which would only cause problems later on because the user doesn't have permissions on it for example.

The PR as it stands, seems to solve the issue for the user that has, exclusively, READ on /.   (a use case that's currently broken).   So it does solve that and has value on its own.

It would be reasonable IMO to tackle the filtering in a separate PR, although, the reason why I initially brought it up is because it looked like all the pieces to do both were right there.  The only thing needed was to not fail if the top level "/" check fails authorization, but give a chance to the subsequent logic below to also look at individual permissions to add in "allowed" scopes.

Basically "/" being a wild card that trumps the rest.

**Comment by EronWright** [comment]
@fpj we need a fix for 0.4 ASAP, thanks

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 1 above)
Comment (fpj) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Ok, let's try to agree on the approach to follow here. The options I see are:

1- Stick to the file system model and keep the change as is. I'd still like to add a test case. but otherwise keep the change as is.
2- Filter according to authorization. It requires a change, but it is a trivial one. I'm on the fence about exposing scope names to all users. It doesn't seem to be a big deal as an unauthorized user can't see the streams in the scope or access stream data.
3- Create an additional call to return only authorized scopes to make it simpler for presentation purposes (e.g., in a UI).

Comment (sarlaccpit) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: I think that leaving "as is" (option 1) is not bad, but is incomplete as it doesn't allow users that only have access to a few scopes, to "list [their] scopes".

Regarding (2), I agree that from a security perspective, it's not a big deal to show scopes they don't have access to because anyway they can't even do a get on these individual scopes or access data.  But, it misguides UIs etc., in allowing users to "select" some scope they may have gotten from the list, and later on have authz issues due to the fact they don't actually have access.

For (3), I feel it would just complicate things to have 2 different calls.  One might say: if I have access on "/" and I use this new filtered scopes list call, what would it show?    Having a list-all and list-just-what-i-can-see feels like it could confuse people (maybe).

To have it all, it seems like option 2, with the filtering, is the way to go.   If a user has READ on /, we return all, if not, we get down to checking individual READ on each scope to produce the filtered list.  (that would be my vote).    Though I do see the appeal of the filesystem analogy, my humble opinion is that the one call that does everything we need trumps it.

Comment (fpj) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Say that for administration purposes, you do want to know all the scopes, would we simply give at least read access to all scopes to such an admin user?

Comment (sarlaccpit) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: I think an admin would be given READ on "/".

Comment (EronWright) [reply controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: TLDR; trim the results of ListScopes.

My full opinion:
1. any resemblance to a file system is superficial; focus on the logical security model that you want.
2. scopes are the foundation for isolation, are likely to have quota or similar policy in the future, and so we should expect the typical user to have access to few scopes.
3. having read access to the admin scopes should be considered a very high level of access
4. it is typical in multi-tenant cloud services that each tenant is 'a world unto themselves'; little-to-no visibility of other tenants.   Enumerating all scopes seems at odds with cloud service principles.

**Comment by tkaitchuck** [comment]
As @fpj mentioned, in a FS when you list a directory like `/` you see all the subdirectories, even the ones you don't have permission to enter and read. Keeping with that analogy, even if you don't have permission to see the streams in a scope that doesn't mean you shouldn't be able to see that the scope itself exists. 

This does not make things worse form a security point of view, because even if the scope wasn't returned from a list call an application would be able to infer its existence by attempting to create a scope with the same name.

**Comment by EronWright** [comment]
On one hand the FS analogy is not helpful.  On the other hand we needed this fixed a week ago and will accept anything.

**Comment by fpj** [comment]
In the interest of time, I'll proceed with the change as proposed. I'll fix the tests and update this PR. Thanks everyone for chiming in.

**Comment by EronWright** [comment]
@tkaitchuck, consider that createScope is a high-privilege operation; most users would get an authorization failure.  Information leakage is bad from a cloud/multi-tenant design perspective.  Not to mention a usability one.

@fpj to recap what will happen, the keycloak plugin will interpret `authorize("/", principal, READ)` as a low-privilege operation.  Meanwhile the UI will show all scopes but produce an 'access denied' message when the user attempts to access the wrong scope.

**Comment by fpj** [comment]
@EronWright I fully agree that it would be odd if not problematic for a cloud service to expose all environments to all existing users. I find it arguable whether this kind of access control and visibility should be implemented at the level of a component like Pravega. A component like the controller should provide the minimum amount of support to enable authentication and authorization, and additional functionality should be implemented on top.  

On the file system analogy, I'm not stuck on the idea of file systems, and I've been using it as a way to illustrate a hierarchical scheme, as opposed to a flat scheme (resources are fully independent and have no hierarchical relation).

My recommendation is that we stick to fixing the bug in this PR, do it shortly, and if there is a desire to revisit our multi-tenancy model, we do it separately.

File: controller/src/test/java/io/pravega/controller/rest/v1/Failing403StreamMetaDataTests.java [hunk 2]
```diff
@@ -17,10 +17,11 @@
 import io.pravega.controller.server.rpc.auth.TestAuthHandler;
 import org.junit.Before;
 
+
```
Comment (shiveshr) [inline controller/src/test/java/io/pravega/controller/rest/v1/Failing403StreamMetaDataTests.java]: please remove this line

File: controller/src/test/java/io/pravega/controller/rest/v1/FailingSecureStreamMetaDataTests.java [hunk 3]
```diff
@@ -15,6 +15,7 @@
 import io.pravega.controller.server.rpc.auth.PravegaAuthManager;
 import io.pravega.controller.server.rpc.grpc.impl.GRPCServerConfigImpl;
 import io.pravega.test.common.TestUtils;
+
```
Comment (shiveshr) [inline controller/src/test/java/io/pravega/controller/rest/v1/FailingSecureStreamMetaDataTests.java]: please remove unnecessary line

File: controller/src/test/java/io/pravega/controller/rest/v1/SecureStreamMetaDataTests.java [hunk 4]
```diff
@@ -41,4 +57,5 @@ public void setup() {
         map.addAll(HttpHeaders.AUTHORIZATION, TestUtils.basicAuthToken("admin", "1111_aaaa"));
         return request.headers(map);
     }
+
```
Comment (shiveshr) [inline controller/src/test/java/io/pravega/controller/rest/v1/SecureStreamMetaDataTests.java]: please remove unnecessary line
